### PR TITLE
Add a Dockerfile to allow easy creation of draft PDFs

### PR DESCRIPTION
**Pull request by tarleb**
Intended use:

    # Build the openjournals/paperdraft image.
    docker build \
           --tag=openjournals/paperdraft \
           --file paperdraft.Dockerfile .

    # Use image to compile a paper.md file into paper.pdf.
    docker run --rm \
           --volume $PWD/paper/folder:/data \
           --user $(id -u):$(id -g) \
           --env JOURNAL=joss \
           openjournals/paperdraft

### CODE REVIEW

**Comment by tarleb** [comment]
I've uploaded a demo image as `tarleb/paperdraft` to Docker Hub. Users have to point to the directory of their paper via `--volume /absolute/path/to/paper:/data`.

**Comment by arfon** [comment]
@tarleb this looks great! One quick question - does this assume the presence of a file named `paper.md` or is the path to the paper passed explicitly?

**Comment by tarleb** [comment]
The `paper.md` input is currently hard-coded and cannot be overwritten. Alternatively, we could either

1. remove any default and require users to specify the input explicitly, or
2. allow to override the input file name.

I opted for the current implementation because that seems to be what whedon does, but it wouldn't be hard to change.

**Comment by arfon** [comment]
Option 2 sounds good to me @tarleb. Am I right in thinking that we're not actually using Whedon under the hood here, instead just passing all of the correct commands/variables to Pandoc to produce the PDFs? In which case, I think we can just ask the author to be explicit about the path/location of the `paper.md`.

**Comment by tarleb** [comment]
Changed it to option 2.

> Am I right in thinking that we're not actually using Whedon under the hood here, instead just passing all of the correct commands/variables to Pandoc to produce the PDFs?

Yes, I figured this would be easier than using (and configuring) full whedon.

**Comment by arfon** [comment]
Apologies @tarleb, but I'm having a little trouble running this (this is partly due to my limited Docker experience).

I've managed to build this Docker image locally (last line of output `Successfully tagged openjournals/paperdraft:latest`).

Now assuming I have a paper on my computer at `/Users/arfon/Sites/fidgit/paper/paper.md` then should the following command work (from path `/Users/arfon/Sites/fidgit`):

```
docker run --rm \
    --volume /Users/arfon/Sites/fidgit/paper/paper.md \
    --user $(id -u):$(id -g) \
    --env JOURNAL=joss \
    openjournals/paperdraft
```

When I run this command I get the following error:

```
pandoc: paper.md: openBinaryFile: does not exist (No such file or directory)
```

Any suggestions as to what I'm doing wrong here?

**Comment by tarleb** [comment]
I think changing `--volume /Users/arfon/Sites/fidgit/paper/paper.md` to `--volume /Users/arfon/Sites/fidgit/paper:/data` should do the trick.
The above tells docker to make the directory `/Users/arfon/Sites/fidgit/paper/` available to Docker under the path `/data`. That directory is set as the image's working directory, so pandoc can find `paper.md` there.

File: paperdraft.Dockerfile [hunk 1]
```diff
@@ -0,0 +1,31 @@
+FROM pandoc/latex:2.9.2.1 AS papergen
+
+# Install additional LaTeX packages
+RUN tlmgr install \
```
Comment (arfon) [inline paperdraft.Dockerfile:4]: @tarleb - not sure if you've seen this list already but here are the packages Whedon currently has available: https://github.com/openjournals/heroku-buildpack-tex/blob/master/bin/compile#L85-L109

**Comment by arfon** [comment]
Thanks @tarleb - that works great!

File: paperdraft.Dockerfile (hunk 1 above)
Comment (tarleb) [reply paperdraft.Dockerfile:4]: Thanks, I hadn't seen that yet. The packages are hard-coded for now, maybe we can find a way to share the list (or the image).

**Comment by tarleb** [comment]
An example GitHub action using the image. 

``` yaml
on:
  push:

jobs:
  joss-article:
    runs-on: ubuntu-18.04
    steps:
      - uses: actions/checkout@v1
      - uses: docker://tarleb/paperdraft:latest
        with:
          args: example/article.md

      - name: Upload
        uses: actions/upload-artifact@v1
        with:
          name: paper
          path: paper.pdf
```

Here it is in action: https://github.com/pandoc-scholar/pandoc-scholar/actions/runs/140699164
The resulting PDF is made available as an artifact. Now checking if we can make this more convenient to use.

**Comment by tarleb** [comment]
I wrote an experimental GitHub Action here: <https://github.com/tarleb/openjournals-draft-action>. It works fine, but doesn't seem to buy us much compared to the docker-only approach:

``` yaml
on: [push]

jobs:
  paper:
    runs-on: ubuntu-latest
    name: Paper Draft
    steps:
      - uses: actions/checkout@v2
      - name: Build draft PDF
        uses: tarleb/openjournals-draft-action@master
        with:
          paper-path: docs/joss/paper.md
      - name: Upload
        uses: actions/upload-artifact@v1
        with:
          name: paper
          path: docs/joss/paper.pdf
```

Demo on an actual JOSS paper: https://github.com/tarleb/ESA/actions/runs/141073499

**Comment by arfon** [comment]
> I wrote an experimental GitHub Action here: https://github.com/tarleb/openjournals-draft-action. It works fine, but doesn't seem to buy us much compared to the docker-only approach:

I think this gets us one big thing - it removes the need for authors to install Docker which I think shouldn't be underestimated as a big plus.

**Comment by tarleb** [comment]
That's true, I wasn't clear. What I meant was the difference between

``` yaml
- uses: docker://tarleb/paperdraft:latest
  with:
    env:
      JOURNAL: joss
    args: example/article.md
```

which just requires the docker image on Docker Hub, and

``` yaml
- uses: tarleb/openjournals-draft-action@master
  with:
    journal: joss
    paper-path: docs/joss/paper.md
```

which additionally requires a separate GitHub Action repository. The latter is a *bit* cleaner and better documented, so it's probably worth it.

**Comment by arfon** [comment]
/ cc @xuanxu for visibility here. It would be good to talk with @tarleb about how we can integrate this with Buffy.

**Comment by arfon** [comment]
@tarleb - I published the Action here: https://github.com/marketplace/actions/open-journals-pdf-generator

**Comment by arfon** [comment]
@tarleb - can we update the CSL here too?

**Comment by tarleb** [comment]
On it; this just requires a rebuild of the docker image with no code changes.

While I'm at it, I'll also do some config tweaks so we can update to the latest pandoc version 2.11.0.4.

**Comment by tarleb** [comment]
Done and done. The new image has been pushed to Docker Hub.

**Comment by arfon** [comment]
> While I'm at it, I'll also do some config tweaks so we can update to the latest pandoc version 2.11.0.4.

BTW, if we update Whedon to Pandoc `v2.11.0.4`, do we need to make changes to the way that we're calling out to Pandoc e.g. here: https://github.com/openjournals/whedon/blob/8db9f43069bdef150a50adf2a1e811d47f35ca59/lib/whedon/compilers.rb#L161-L173

**Comment by tarleb** [comment]
The only necessary change will be to replace `--filter pandoc-citeproc` with `--citeproc`, as citation processing no longer requires a filter but is now fully baked into pandoc. I believe John will release a new version 2.11.1 within the next few days. It will also include a few citation-related bugfixes. Probably best to delay updating till next week.
The changes for Docker took a bit longer because they uncovered a bug in pandoc's YAML parsing (fixed it in the dev version).

**Comment by arfon** [comment]
Got it. Thanks @tarleb. With https://github.com/openjournals/whedon-api/pull/119, I've just updated Whedon for JOSS to use Pandoc `v.2.11.1` so if we can get this Dockerfile updated too then I then we'll have them both in sync and we can start more widely promoting this action to JOSS authors.

**Comment by arfon** [comment]
@tarleb - just checking - when you've been pushing `openjournals/openjournals-draft-action` to Dockerhub, has that being built using the Dockerfile in this repo?

I'm trying to work out if we need to update the [GitHub Action](https://github.com/marketplace/actions/open-journals-pdf-generator) to point to `openjournals/paperdraft` (as defined [here](https://github.com/openjournals/whedon/pull/74/files#diff-dbade1e5797a7633e024a7685a43a7cbd96ae6fefc6314294807ae34cd83712cR12)) once we get the build step running here (it's failing on auth for push right now I think https://github.com/openjournals/whedon/actions/runs/377055010)

**Comment by tarleb** [comment]
Yes, that's correct. I did pretty much exactly what the GitHub Action is trying to do.

**Comment by tarleb** [comment]
I think the problem with the login step is a missing backslash `\` at the end of the `echo` line.

**Pull request merged**